### DISCUSSION TITLE
Document's relevancy score weight

### DISCUSSION THREAD

**Discussion by Jure-BB** [meilisearch/product#216] [Feedback & Feature Proposal]
An option to specify a numeric field as document's relevancy score weight. 

At search time, document's relevancy scores (i.e. ranking rules) would be multiplied with document's weight value. Weight value used as multiplier only influences sort order by nudging relevancy and is not a regular hard sort rule.

**Comment by gmourier** [comment]
Hi @Jure-BB 👋,

Today the order of searchable attributes represents a kind of general weight applied by field to all documents. If I understand correctly you would like to have a different weight for each document, is that right?

Can you tell me more about your use case, I imagine that not all documents are equivalent in terms of information and that some deserve to be more important according to your own definitions.

Thanks for your feedback

**Reply by Jure-BB** [reply]
Yes, that's right. 

I was evaluating MeiliSearch for a geocoder use case. For example: relevancy of a settlement is affected by population size and road relevancy is affected by road rank. There are also other document types like full addresses and named places. While sorting by order of attributes is possible, fine tweaking of a relevancy that combines distance of a place from user, multiple ranking attributes and input typos can be difficult.

**Comment by slava-vishnyakov** [comment]
Actually, one more use case would be "newer/fresh documents have higher weights", so that you could manually add a weight of 1.0 to 0.5 where new documents would get a boost.

**Comment by braco** [comment]
Really surprised boost/penalize by arbitrary fields isn't already developed – can't imagine industrial search without this.

https://www.algolia.com/doc/guides/managing-results/must-do/custom-ranking/how-to/boost-or-penalize-some-records/
https://www.elastic.co/guide/en/elasticsearch/reference/7.17/mapping-boost.html

The importance of some fields vs others is also a pretty critical feature. (Match in title is more important than match in tertiary attribute)

**Reply by irevoire** [reply]
Hey, aren’t our ranking rules doing the exact same thing as the « boosted » ranking rule of algolia?
All the examples I see on the page you shared can be implemented with almost exactly the same code in meilisearch (i.e.: `boosted:desc`)

**Reply by braco** [reply]
Hey @irevoire. I think your ranking rules are based on ordinals and not more specific weights that can be fine tuned at query time. Right?

https://github.com/meilisearch/product/discussions/270

The value-based use cases include number fields like upvotes, dates, number of citations, etc, that would make a document more relevant to the user, but how much more relevant those are is something that needs to be fine-tuned in almost all cases.

**Reply by braco** [reply]
FWIW this is listed as a differentiator on Typesense's comparison page:

https://typesense.org/typesense-vs-algolia-vs-elasticsearch-vs-meilisearch/

> "Query Field Weights & Boosting"

**Comment by sandstrom** [comment]
Just a note, this functionality is probably a subset of [function scoring](https://github.com/meilisearch/product/discussions/205).

**Reply by gmourier** [reply]
Hey @sandstrom,

Thanks for mentioning it! 

It could also be somewhat related to [that](https://github.com/orgs/meilisearch/discussions/648) feature.

**Reply by sandstrom** [reply]
@gmourier Thanks!

If I were you, I'd take my inspiration from ElasticSearch. Their function scoring[1] is more powerful than Algolia filter scoring.

Also, I know that you have function scoring on your roadmap, and that it hasn't gotten a lot of votes. But people don't always know what they want.

For example, the following items from your roadmap would all be solved by function scoring:

- "Rank/Sort with "near""
- "Scored Filter"
- "Field Weighting"
- "Promoting documents"

But people looking at the roadmap doesn't know that, so they'll vote for those individual items, instead of function scoring which would solve all of them. But when you as a PM think about the roadmap, sometimes it's better to build one slightly larger feature that solves 10+ "smaller requests", than building 10 smaller features to solve individual problems.

I know function scoring (roughly defined as it's implemented in ElasticSearch, though obviously you should do your own 'take' on it) is some work. But it's also one of those foundational features that will solve a lot of "long-tail problems".

[1] https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html
